### PR TITLE
[estree] make function declaration id nullable

### PR DESCRIPTION
**Pull request by stas-vilchik**
A `FunctionDeclaration` `id` can be `null` if this function declaration is a part of the `ExportDefaultDeclaration` ([ast explorer gist](https://astexplorer.net/#/gist/bcb2d30873c46351a69473ad91a9cadd/2ccd96c9b4202b442c6f7fae52acc14cd5fca756)):

```javascript
export default function() {}
```

This PR updates the `FunctionDeclaration` interface.

### CODE REVIEW

**Comment by RReverser** [comment]
For compatibility with existing code and better typecheck, I'd prefer this to be forked into separate interface which could only appear in export.

**Comment by stas-vilchik** [comment]
@RReverser I'm not sure what could be the best way to do that. If I created a new interface with nullable `id`:

```typescript
export interface ExportDefaultFunctionDeclaration extends BaseFunction, BaseDeclaration {
  type: "FunctionDeclaration";
  id: Identifier | null;
  body: BlockStatement;
}
```

it would share the same `type` as the normal `FunctionDeclaration`. Later if I wanted to create an `eslint` rule using this typings, I would write something like:

```typescript
const rule: Rule.RuleModule = {
  create(context: Rule.RuleContext) {
    return {
      FunctionDeclaration(node: Node) {
        (node as FunctionDeclaration).id; // nullable!
      },
```

In this case I'd need to remember (or even know) that I can't safely cast `node` to `FunctionDeclaration`, instead I should cast it to `FunctionDeclaration | ExportDefaultFunctionDeclaration`. I think this approach is quite error-prone.

File: types/estree/index.d.ts [hunk 1]
```diff
@@ -196,7 +196,8 @@ interface BaseDeclaration extends BaseStatement { }
 
 export interface FunctionDeclaration extends BaseFunction, BaseDeclaration {
   type: "FunctionDeclaration";
-  id: Identifier;
+  // `id` is null when a function declaration is a part of the `export default function` statement
```
Comment (Kovensky) [inline types/estree/index.d.ts]: This could be a JSDoc comment

File: types/estree/estree-tests.ts [hunk 2]
```diff
@@ -223,7 +223,7 @@ boolean = memberExpression.computed;
 
 // Declarations
 var functionDeclaration: ESTree.FunctionDeclaration;
-identifier = functionDeclaration.id;
+var identifierOrNull: ESTree.Identifier | null = functionDeclaration.id;
```
Comment (Kovensky) [inline types/estree/estree-tests.ts]: `functionDeclaration.id = null` would be a better test, as the `identifierOrNull` test wouldn't fail even without the change to the declaration.

**Comment by RReverser** [comment]
Not sure what is `RuleModule` in your snippet, but if it's kind of a visitor, it should be redefined to accept the broader `FunctionDeclaration` variant with nullable `id`, while the non-nullable one would still be used for statement level.

**Comment by Kovensky** [comment]
The only way to do that would be to know what the type of the parent node is, and that is not part of the type of a node.

Even if you do add another interface, say, `NonNullableFunctionDeclaration`, the moment you `|` it with `FunctionDeclaration` inside the definition of `Declaration`, it'd make `FunctionDeclaration` widen to be the exact same type as `NonNullableFunctionDeclaration`.

By the way, while we're at it, I just noticed that `ClassDeclaration` also has non-nullable `id` but it _also_ is nullable if it's the `export default`.

**Comment by stas-vilchik** [comment]
@RReverser my bad, I should have mentioned that it was an example of eslint rule. Its visitor relies on the node type and [returns a generic](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint/index.d.ts#L254-L259) `Node`. Therefore as a custom rule author, I'd need to remember that the `FunctionDeclaration` visitor doesn't simply return a `estree.FunctionDeclaration`, but something different.

---

@Kovensky thanks for the review comments and for the `ClassDeclaration` case. I'm not sure to understand your idea though. 

Anyway I think it'd better if the `estree.FunctionDeclaration` represented a node with the type `FunctionDeclaration` with all possible variants.

**Comment by RReverser** [comment]
I surely understand the complexities of updating this in a clean way, but want to find a way that would break as little code out there as possible.

**Comment by stas-vilchik** [comment]
@RReverser I understand, I just don't know how to achieve that.

**Comment by RyanCavanaugh** [comment]
The TypeScript syntax tree suffers the same problem and it's a big source of bugs.

I think the fix as-is is honestly the right one - attempting to split out the tree differently doesn't really work because it's really a matrix (export, default, named, async, etc) rather than a directable graph.

@RReverser without a concrete better PR I think this one should be merged. Any objections?

**Comment by RReverser** [comment]
Yeah, I think at this point trying to split out turns out to be too complicated to be worth it. We can always try to improve AST in future. @stas-vilchik sorry for the delay on this one!

**Review by RReverser — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
🌟 🎈 🎉 🏆 🎂 ✨ ⭐️ 

Congratulations on your first DefinitelyTyped contribution!

🌟 🎈 🎉 🏆 🎂 ✨ ⭐️

**Comment by stas-vilchik** [comment]
Great, thanks!

**Pull request merged**